keg_auth/model/__init__.py

```diff
@@ -0,0 +1,521 @@
+    )
+    if rel_property:
+        setattr(
+            user_cls, rel_property,
+            sa.orm.relationship(bundle_cls, secondary=table, backref='users')
+        )
+    return table
+
+
+def user_group_mapping(user_cls, group_cls, table_name='user_groups',
+                       user_id_attr='id', group_id_attr='id',
+                       rel_property='groups'):
+    table = _make_mapping_table(
+        table_name,
+        user_id=getattr(user_cls, user_id_attr),
+        group_id=getattr(group_cls, group_id_attr),
+    )
+    if rel_property:
+        setattr(
+            user_cls, rel_property,
+            sa.orm.relationship(group_cls, secondary=table, backref='users')
+        )
+    return table
+
+
+def group_permission_mapping(group_cls, permission_cls, table_name='group_permissions',
+                             group_id_attr='id', permission_id_attr='id',
+                             rel_property='permissions'):
+    table = _make_mapping_table(
+        table_name,
+        group_id=getattr(group_cls, group_id_attr),
+        permission_id=getattr(permission_cls, permission_id_attr),
+    )
+    if rel_property:
+        setattr(
+            group_cls, rel_property,
+            sa.orm.relationship(permission_cls, secondary=table)
+        )
+    return table
+
+
+def group_bundle_mapping(group_cls, bundle_cls, table_name='group_bundles',
+                         group_id_attr='id', bundle_id_attr='id',
+                         rel_property='bundles'):
+    table = _make_mapping_table(
+        table_name,
+        group_id=getattr(group_cls, group_id_attr),
+        bundle_id=getattr(bundle_cls, bundle_id_attr),
+    )
+    if rel_property:
+        setattr(
+            group_cls, rel_property,
+            sa.orm.relationship(bundle_cls, secondary=table, backref='groups')
+        )
+    return table
+
+
+def initialize_mappings(namespace='keg_auth', registry=entity_registry.registry):
+    def _make_table_name(default_name):
+        return '{}_{}'.format(namespace, default_name) if namespace else default_name
+
+    mappings = {
+        'user_permissions': (user_permission_mapping, 'user', 'permission'),
+        'bundle_permissions': (bundle_permission_mapping, 'bundle', 'permission'),
+        'user_bundles': (user_bundle_mapping, 'user', 'bundle'),
+        'user_groups': (user_group_mapping, 'user', 'group'),
+        'group_permissions': (group_permission_mapping, 'group', 'permission'),
+        'group_bundles': (group_bundle_mapping, 'group', 'bundle')
+    }
+    tables = {}
+    for base_name, mapping_data in mappings.items():
+        table_func, type1, type2 = mapping_data
+        table1 = registry.get_entity_cls(type1)
+        table2 = registry.get_entity_cls(type2)
+
+        tables[base_name] = table_func(table1, table2, table_name=_make_table_name(base_name))
+    return tables
+
+
+def initialize_events(registry=entity_registry.registry):
```

Comment:
I don't think this will become an issue. It is important to invalidate sessions only when necessary. Performance is less important in this case, because user ops like this, changing permissions around, etc. are not very common operations (compared to flushes during ETL processes with thousands of records, for example).

Reply:
yea, but in tests this is a big deal. This code checks for a changed user, group and bundle for every new ORM object... that's a lot of `isinstance` calls in our tests.

Reply:
Fair enough, you made me look. The impact was not very bad, but I have removed `isinstance` usage in favor of a more strict type comparison. That does make a small difference when we ramp up the number of objects into the thousands (and the difference was already pretty small). The performance hit for having these events in place is negligible.

FWIW, I would have rathered put the event handlers on the objects themselves rather than in `before_flush`, but SA doesn't support doing operations on other [related] objects in that context.